src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+    return observer;
+  }
+
+  struct EntryCallbacks final {
+    const HashableV8Ref<v8::Module>& keyForRow(const Entry& entry) const {
+      return entry.key;
+    }
+    bool matches(const Entry& entry, const HashableV8Ref<v8::Module>& key) const {
+      return entry.key == key;
+    }
+    uint hashCode(const HashableV8Ref<v8::Module>& ref) const {
+      return ref.hashCode();
+    }
+  };
+
+  struct ContextCallbacks final {
+    const SpecifierContext& keyForRow(const Entry& entry) const {
+      return entry.specifier;
+    }
+    bool matches(const Entry& entry, const SpecifierContext& specifier) const {
+      return entry.specifier == specifier;
+    }
+    uint hashCode(const SpecifierContext& specifier) const {
+      return specifier.hashCode();
+    }
+  };
+
+  struct UrlCallbacks final {
+    const Url& keyForRow(const Entry& entry) const {
+      return entry.specifier.specifier;
+    }
+    bool matches(const Entry& entry, const Url& specifier) const {
+      return entry.specifier.specifier == specifier;
+    }
+    uint hashCode(const Url& specifier) const {
+      return specifier.hashCode();
+    }
+  };
+
+  // Resolves the module from the inner ModuleRegistry, caching the results.
+  kj::Maybe<Entry&> resolveWithCaching(Lock& js, const ResolveContext& context)
+      KJ_WARN_UNUSED_RESULT {
+    ResolveContext innerContext {
+      // The type identifies the resolution context as a bundle, builtin, or builtin-only.
+      .type = context.type,
+      // The source identifies the method of resolution (static import, dynamic import, etc).
+      // This is passed along for informational purposes only.
+      .source = context.source,
+      // The inner registry should ignore all URL query parameters and fragments
+      .specifier = context.specifier.clone(Url::EquivalenceOption::IGNORE_FRAGMENTS |
+                                           Url::EquivalenceOption::IGNORE_SEARCH),
+      // The referrer is passed along for informational purposes only.
+      .referrer = context.referrer,
+    };
+    KJ_IF_SOME(found, inner.resolve(innerContext)) {
+      return kj::Maybe<Entry&>(lookupCache.upsert(Entry {
+        .key = HashableV8Ref<v8::Module>(js.v8Isolate,
+            check(found.getDescriptor(js, getObserver()))),
+        // Note that we cache specifically with the passed in context and not the
+        // innerContext that was created. This is because we want to use the original
+        // specifier URL (with query parameters and fragments) as part of the key for
+        // the lookup cache.
+        .specifier = context,
+        .module = found,
+      }, [](auto&,auto&&) {}));
+    }
+    return kj::none;
+  }
+
+  kj::Table<Entry, kj::HashIndex<EntryCallbacks>,
+                   kj::HashIndex<ContextCallbacks>,
+                   kj::HashIndex<UrlCallbacks>> lookupCache;
+  friend class SyntheticModule;
+};
+
+v8::MaybeLocal<v8::Value> SyntheticModule::evaluationSteps(
```

Comment:
Nit: Would moving evaluationSteps into the class definition be cleaner? This can stay as-is if you think having it around the other big functions here makes more sense.

Reply:
It's here because the implementation depends on the definition of `IsolateModuleRegistry` which comes after `SyntheticModule`